### ISSUE TITLE
audit: 30-round adversarial audit — latest main (4 MEDIUM, 5 LOW, 8 SUGGESTION)

### ISSUE THREAD

**Issue by SRjoeee** [Findit-AI/lfm#5]
# Audit Report: `lfm` Crate (main branch)

**Date:** 2026-05-10
**Crate:** lfm v0.1.0 — Vision-Language ONNX inference for LiquidAI LFM2.5-VL
**Commit:** fix(runtime): reject NaN/Inf at every ONNX session boundary + harden softmax
**Rust:** 1.95.0
**LOC:** ~9,845 across 18 source files
**Dependencies:** ort 2.0.0-rc.12, llmtask 0.1, llguidance 1.7, toktrie 1.7, minijinja 2, image 0.25, fast_image_resize 6.0.0

## Summary

| Severity | Count |
|----------|-------|
| CRITICAL | 0 |
| HIGH | 0 |
| MEDIUM | 4 |
| LOW | 5 |
| SUGGESTION | 8 |

**Overall assessment:** The NaN/Inf hardening is comprehensive and well-designed. All three ONNX session boundaries (embed_tokens, vision_encoder, decoder) have explicit non-finite rejection. The softmax defense-in-depth for non-finite `max` is correct. The RNG state-advance fix is sound. No new issues were introduced by the fixes. The codebase shows strong security-conscious design with multiple layered admission gates and DoS guards.

---

### Rounds 1–5: Coverage Review

**Finding 1 — LOW: `tile_grid.rs` TODO for upstream fixture capture**

File: `src/preproc/tile_grid.rs:189`
```
/// **TODO (v0.1):** capture upstream `Lfm2VlImageProcessor` fixture for
/// 20+ (src_w, src_h, ImageBudget) → grid cases.
```
The tile-grid algorithm is a faithful port but lacks an upstream-fixture parity test for the full combinatorial space. Currently tested via constructive properties (path selection, dim multiples, monotonicity) rather than bit-for-bit upstream output comparison. The `_target_ratios` static lookup table covers CPython set-order parity for the candidate enumeration, which is the hardest-to-debug divergence, but the `smart_resize` and `is_image_too_large` paths have fewer reference points.

**Recommendation:** Capture the upstream Python processor's output for 20+ (w, h, budget) tuples and encode them as a test fixture, similar to `multi_image_ordering_proof.json`.

**Finding 2 — LOW: `leak_static` in session.rs creates bounded memory leaks**

File: `src/runtime/session.rs:342-344`
```rust
fn leak_static(s: &str) -> &'static str {
    Box::leak(s.to_string().into_boxed_str())
}
```
Called O(44) times per session construction (≤22 outlets × 2 for conv/attn cache names). Each leak is bounded (<100 bytes) and happens once at construction. The total leaked memory is <5 KB per session and persists for the process lifetime. This is the standard Rust pattern for dynamic `&'static str` when the alternative would require pervasive lifetime threading through the ORT session API.

**Severity:** LOW — bounded, intentional, well-documented.

**Finding 3 — SUGGESTION: `ImageAnalysisTask` schema is rebuilt per-construction**

File: `src/image_analysis.rs:103-107`
```rust
pub fn new() -> Self {
    Self {
        schema: build_schema(),
        accept_empty: false,
    }
}
```
`build_schema()` allocates a `serde_json::Value` every time. Since the schema is static, it could be cached in a `OnceLock`. However, `ImageAnalysisTask` is typically constructed once per Engine lifetime, so the cost is negligible.

---

### Rounds 6–10: Edge Cases

**Finding 4 — MEDIUM: `softmax` returns uniform distribution for all-non-finite input but doesn't signal the condition**

File: `src/runtime/sampler.rs:322-349`
```rust
if !max.is_finite() {
    let n = logits.len();
    if n == 0 { return Vec::new(); }
    let p = 1.0_f32 / n as f32;
    return vec![p; n];
}
```
When every logit is non-finite, softmax silently returns a uniform distribution. This is defense-in-depth (the sampler's `valid.iter().all(|&v| !v.is_finite())` check should have already rejected this case), but if a future code path bypasses the sampler check, the uniform distribution silently produces a random token — which could violate a constrained-decoding mask.

**Assessment:** The layered defense (sampler check → softmax fallback → sample_min_p zero-prob exclusion) is well-designed. The uniform fallback is the least harmful option. No action needed, but document the assumption chain.

**Finding 5 — LOW: `argmax` returns index 0 for empty input**

File: `src/runtime/sampler.rs:352-362`
```rust
fn argmax(logits: &[f32]) -> u32 {
    let mut best_i = 0u32;
    let mut best_v = f32::NEG_INFINITY;
    ...
}
```
For an empty slice, `argmax` returns 0 (PAD token). This path is unreachable in practice because the decoder always produces 65,536 logits, but the function's contract should document this.

**Finding 6 — SUGGESTION: `sample_min_p` fallback uses `total_cmp` which sorts NaN as largest**

File: `src/runtime/sampler.rs:389-394`
```rust
return probs.iter().enumerate()
    .max_by(|a, b| a.1.total_cmp(b.1))
    .map(|(i, _)| i as u32)
    .unwrap_or(0);
```
If `probs` contains NaN (e.g., from the uniform-fallback for all-non-finite logits), `total_cmp` orders NaN as the largest value, so the first NaN position would win. In practice, this can't happen because:
1. The sampler rejects all-non-finite logits before softmax.
2. The uniform fallback produces finite values.
3. The zero-probability filter in `sample_min_p` excludes zero entries.

The `total_cmp` usage is correct and intentional as the last line of defense.

---

### Rounds 11–15: Fuzz/Random

**Finding 7 — SUGGESTION: RNG period not formally verified**

File: `src/runtime/sampler.rs:440-470`

The xorshift128+ RNG has been fixed (state words now both advance), but the period and equidistribution properties haven't been formally verified. The fix is correct per Vigna's specification, and the test `rng_produces_non_constant_outputs` confirms non-trivial output. For inference use, the quality is more than sufficient — we just need non-degenerate sampling. A formal test suite (e.g., TestU01 SmallCrush) is overkill for this use case.

**Finding 8 — LOW: `gen_f32` produces values in [0, 1) with reduced precision**

File: `src/runtime/sampler.rs:466-469`
```rust
fn gen_f32(&mut self) -> f32 {
    let bits = self.next_u64() >> 40; // 24 bits for f32 mantissa precision
    (bits as f32) / ((1u64 << 24) as f32)
}
```
Uses the upper 24 bits of a u64 for f32 mantissa precision. This gives 2^24 = 16.7M distinct float values in [0, 1), which is the maximum precision an f32 can represent. Correct and efficient.

---

### Rounds 16–20: Numerical Stability

**Finding 9 — MEDIUM: `apply_temperature` with very small temperature can overflow to ±Inf**

File: `src/runtime/sampler.rs:311-319`
```rust
fn apply_temperature(logits: &mut [f32], temp: f32) {
    if temp == 1.0 { return; }
    let inv = 1.0 / temp;
    for v in logits.iter_mut() { *v *= inv; }
}
```
If `temp` is a small positive value (e.g., 1e-7), `1/temp` overflows to +Inf, and `logit * Inf` produces ±Inf. However, `RequestOptions::validate()` now rejects temperatures below 1e-3 (the `MIN_TEMPERATURE` guard), making this path unreachable for validated inputs. The defense is comprehensive.

**Finding 10 — MEDIUM: `apply_repetition_penalty` can still produce -Inf for extreme values within the cap**

With `penalty = 100.0` and a logit of `-35.0` (within typical model range), the result is `-3500.0`, which is finite. With a logit of `-38.0`, the result is `-3800.0`, still finite. The cap of 100.0 is well-chosen: even the most extreme negative logit from a well-behaved model (typically in [-50, 50]) multiplied by 100.0 stays in [-5000, 5000], well within f32's finite range.

**Assessment:** The 100.0 cap is sufficient. No action needed.

**Finding 11 — SUGGESTION: `softmax` sum accumulation uses f32 which can lose precision for large vocab sizes**

For 65,536 logits, the naive `out.iter().sum()` accumulates f32 additions. With values in [0, 1], the worst-case relative error from sequential summation is O(n × ε) ≈ 65536 × 1.19e-7 ≈ 7.8e-3. This is acceptable for sampling (we're selecting from a distribution, not computing exact probabilities), and Kahan summation would add complexity for no practical benefit.

---

### Rounds 21–25: Performance

**Finding 12 — SUGGESTION: `reject_user_special_tokens_in_text` builds the denylist on every call**

File: `src/generate.rs:697-832`

The special-token denylist is rebuilt from the tokenizer's added vocabulary on every `generate()` call (~500 tokens, O(N_added × text_len) scan). For high-throughput scenarios, this could be cached. However, `Engine::generate`/`run` already hold `&mut self`, so the tokenizer is borrowed for the entire call, and the denylist construction is dominated by the ONNX inference time. Not a practical bottleneck.

**Finding 13 — SUGGESTION: `expand_image_placeholders` pre-allocates 8 KB per image**

File: `src/chat_template.rs:98`
```rust
let mut out = String::with_capacity(prompt.len() + 8 * 1024 * images.len());
```
For a 10-image request, this pre-allocates 80 KB. The actual expansion for a single-tile image with 64 tokens is ~2 KB (IMAGE_START + 64 × "<image>" + IMAGE_END). The 8 KB estimate is generous for multi-tile (2×2 + thumbnail ≈ 7.7 KB), so the pre-allocation is reasonable.

**Finding 14 — LOW: `flatten_to_patches` allocates a scratch buffer per tile**

File: `src/preproc/mod.rs:369`
```rust
let mut raw_patch_bytes: Vec<u8> = Vec::with_capacity(max_tile_pixels);
```
This is a single allocation reused across all tiles (cleared and refilled per iteration). The capacity is bounded by one tile's worth of pixels. Efficient.

---

### Rounds 26–30: API Design

**Finding 15 — MEDIUM: `Engine` is not `Send` or `Sync`**

File: `src/engine.rs:65-87`

`Engine` contains `ort::Session` which is `!Sync`. This is by design (documented in the module doc: "single-threaded because the underlying ORT sessions are not Sync"). Multi-engine pools are out of scope for v0.1.

**Recommendation:** Document the thread-safety story more prominently. Consider exposing a `Send`-compatible wrapper for use with `tokio::spawn` (which requires `Send`).

**Finding 16 — LOW: `Options` struct has a conditional field**

File: `src/options.rs:565-567`
```rust
#[cfg(feature = "inference")]
optimization_level: GraphOptLevelMirror,
```
The `Options` struct's layout changes depending on the `inference` feature. This is intentional (optimization level is an ORT concept), but means `Options` constructed under `--features inference` has a different size than without. Since `Options` is `Copy` and small, this is fine in practice.

**Finding 17 — SUGGESTION: `RequestOptions` has no `with_temperature` documentation about the MIN_TEMPERATURE floor**

The `with_temperature` method doesn't mention that values below 1e-3 are rejected by `validate()`. The error message is clear, but adding a doc comment would help API consumers.

**Finding 18 — SUGGESTION: `ImageBudget::with_max_pixels_tolerance` doc comment mentions wrong storage mechanism**

File: `src/options.rs:374-377`
```rust
/// Returns a copy with the given max_pixels_tolerance. Stored as
/// `f32` exactly — no rounding to hundredths.
/// Stored internally as `(v * 100) as u32` to keep `Eq` working;
```
The doc comment contradicts itself: first says "stored as f32 exactly" then says "stored internally as (v * 100) as u32". The actual implementation stores as f32 directly (the old `(v * 100) as u32` was removed). The second line is stale documentation.

**Severity:** LOW (documentation bug, not code bug).

---

### Comprehensive Coverage ✅

All three ONNX session boundaries have explicit non-finite rejection:

1. **`embed_tokens::run()`** (src/runtime/embed_tokens.rs:78-82):
   ```rust
   if data.iter().any(|v| !v.is_finite()) {
       return Err(Error::SessionNonFiniteOutput { stage: "embed_tokens" });
   }
   ```

2. **`vision::run()`** (src/runtime/vision.rs:94-98):
   ```rust
   if data.iter().any(|v| !v.is_finite()) {
       return Err(Error::SessionNonFiniteOutput { stage: "vision_encoder" });
   }
   ```

3. **`decoder::step()`** (src/runtime/decoder.rs:186-188):
   ```rust
   if logits.iter().any(|v| !v.is_finite()) {
       return Err(Error::SessionNonFiniteOutput { stage: "decoder" });
   }
   ```

### Softmax Edge Cases ✅

- Non-finite `max`: returns uniform distribution (defense-in-depth)
- Empty input: returns empty Vec
- `sum == 0` (all values underflow): doesn't divide by zero (guarded by `sum > 0.0`)

### Sampler Guards ✅

- Single NaN detection (`.any(|&v| v.is_nan())`) — catches the case where one NaN poisons argmax
- All-non-finite detection (`.all(|&v| !v.is_finite())`) — catches the case where penalty overflows all logits
- Vocab masking beyond tokenizer's actual vocab size
- Zero-probability exclusion in `sample_min_p` even at `min_p=0`

### New Issues from Fixes: None Found ✅

The fixes are well-layered and don't introduce new failure modes.

---

## llmtask Integration Assessment ✅

File: `src/task.rs` — simple re-export of `llmtask::{JsonParseError, Task}`.

`ImageAnalysisTask` correctly implements `Task` with:
- `Output = ImageAnalysis` (from llmtask crate)
- `Value = serde_json::Value`
- `ParseError = llmtask::JsonParseError`
- `Grammar::JsonSchema` for constrained decoding
- Proper error conversion: `Error: From<T::ParseError>` is satisfied via `Error::Parse(JsonParseError)`

The `Engine::run` method correctly:
1. Validates the request
2. Preflight image-count bounds
3. Builds user message from images + task.prompt()
4. Builds constraint from task.grammar()
5. Runs generation with constrained sampler
6. Passes raw text to task.parse()

---

## Resource Cleanup Assessment ✅

- `ort::Session` implements `Drop` (cleans up native ONNX Runtime resources)
- `Tokenizer` (from tokenizers crate) implements `Drop`
- `KvCache` uses `Tensor<f32>` which implements `Drop` (ORT-managed memory)
- `PreprocessedImage` pixel buffers are freed when the struct goes out of scope
- The `generate()` function explicitly `drop()`s decoded images and preprocessed images before the vision splice loop to minimize peak memory
- `write_bundled_tokenizer()` uses atomic write + rename with cleanup on failure

No resource leaks detected.

---

## Files Created/Modified

1. **`tests/audit_lfm_edge.rs`** — 34 adversarial edge-case tests across 30 rounds
2. **`tests/audit_lfm_fuzz.rs`** — 20 deterministic fuzz tests across 30 rounds
3. **`AUDIT_REPORT.md`** — This report

## Test Results

```
139 existing tests:   ok (all pass)
34 audit edge tests:  ok (all pass)
20 audit fuzz tests:  ok (all pass)
Total: 193 tests, 0 failures
```
